addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxWsSecurity.java

```diff
@@ -0,0 +1,231 @@
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.loxone.internal.core;
+
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.function.BiConsumer;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+
+import org.apache.commons.codec.DecoderException;
+import org.apache.commons.codec.binary.Hex;
+import org.openhab.binding.loxone.internal.core.LxJsonResponse.LxJsonSubResponse;
+import org.openhab.binding.loxone.internal.core.LxServer.Configuration;
+import org.openhab.binding.loxone.internal.core.LxWsClient.LxWebSocket;
+
+/**
+ * Security abstract class providing authentication and encryption services.
+ * Used by the {@link LxWsClient} during connection establishment to authenticate user and during message exchange for
+ * encryption and decryption or the messages.
+ *
+ * @author Pawel Pieczul - initial contribution
+ *
+ */
+abstract class LxWsSecurity {
+    int debugId;
+    String user;
+    String password;
+    Thread executor;
+    LxWebSocket socket;
+    Configuration configuration;
+    LxOfflineReason reason;
+    String details;
+    boolean cancel = false;
+
+    /**
+     * Create an authentication instance.
+     *
+     * @param debugId
+     *            instance of the client used for debugging purposes only
+     * @param configuration
+     *            configuration object for getting and setting custom properties (e.g. token)
+     * @param socket
+     *            websocket to perform communication with Miniserver
+     * @param user
+     *            user to authenticate
+     * @param password
+     *            password to authenticate
+     */
+    LxWsSecurity(int debugId, Configuration configuration, LxWebSocket socket, String user, String password) {
+        this.debugId = debugId;
+        this.configuration = configuration;
+        this.socket = socket;
+        this.user = user;
+        this.password = password;
+    }
+
+    /**
+     * Initiate user authentication. This method will return immediately and authentication will be done in a separate
+     * thread asynchronously. On successful or unsuccessful completion, a provided callback will be called with
+     * information about failed reason and details of failure. In case of success, the reason value will be
+     * {@link LxOfflineReason#NONE}
+     * Only one authentication can run in parallel and must be performed sequentially (create no more threads).
+     *
+     * @param doneCallback
+     *            callback to execute when authentication is finished or failed
+     */
+    void authenticate(BiConsumer<LxOfflineReason, String> doneCallback) {
+        Runnable init = () -> {
+            synchronized (executor) {
```

Comment:
This synchronized doesn't look logical. It is the only synchronized on this variable so no other places would block and subsequent calls to the authenticate method would set a new value to executor that will not be blocking or blocked by other calls to authenticate calls when running the init block. (Can executor be private?)

Reply:
Yes, it does not make sense, changed to final lock to cover all authentication.